File: cpp/src/step_by_step_bench.cpp
/// Step-by-step function benchmark — tests every function in the EmplaceTableWithDecode pipeline.
/// Each function is isolated with real data matching the 4str_0int workload.
///
/// Usage: ./cpp_step_by_step <sel> [iterations]
/// Default: sel=0.1, 10 iterations, 1M rows per iteration

#include <cstdio>
#include <cstdint>
#include <cstring>
#include <cstdlib>
#include <chrono>
#include <vector>
#include <string>
#include <numeric>
#include <algorithm>
#include <random>

#define XXH_INLINE_ALL
#include "xxhash.h"
#include "column_marshaller.h"

static constexpr size_t NUM_STR_COLS = 4;
static size_t G_HT_SIZE = 16384;
static constexpr size_t NUM_PROBE_ROWS = 1000000;
static constexpr size_t BATCH_SIZE = 410;
static constexpr uint64_t SEED = 42;
static constexpr size_t DEFAULT_ITERS = 10;

using Clock = std::chrono::high_resolution_clock;
static double elapsed_ms(Clock::time_point t0, Clock::time_point t1) {
    return std::chrono::duration<double, std::milli>(t1 - t0).count();
}
static inline uint64_t HB(const uint8_t* d, size_t l, uint64_t s) { return XXH3_64bits_withSeed(d, l, s); }

// ═══════════════════════════════════════════════════════════════════
// Data generation
// ═══════════════════════════════════════════════════════════════════

struct TestData {
    std::vector<std::vector<std::vector<uint8_t>>> strCols;
    std::vector<std::vector<taper::VarcharSlice>> slices;
    std::vector<int64_t> hashes;
    std::vector<int64_t> values;
    size_t totalRows;
    size_t numKeys;
    size_t numChunks;
};

static TestData GenData(double sel) {
    // ─── Two params only: G_HT_SIZE + sel ───
    size_t numChunks = G_HT_SIZE / 8;
    if (numChunks < 1) numChunks = 1;
    size_t nc = 1; while (nc < numChunks) nc <<= 1; numChunks = nc;
    size_t capacity = numChunks * 8;

    size_t distinctKeys = static_cast<size_t>(capacity * 0.89);
    if (distinctKeys < 1) distinctKeys = 1;

    size_t numKeys = static_cast<size_t>(distinctKeys * sel);
    if (numKeys < 1) numKeys = 1;
    size_t probeMisses = distinctKeys - numKeys;
    size_t probeHits = (NUM_PROBE_ROWS > probeMisses) ? (NUM_PROBE_ROWS - probeMisses) : 0;
    size_t numProbeRows = probeHits + probeMisses;

    fprintf(stderr, "  sizing: numChunks=%zu, capacity=%zu, distinctKeys=%zu\n", numChunks, capacity, distinctKeys);
    fprintf(stderr, "  build: numKeys=%zu (%.1f%% of capacity)\n", numKeys, 100.0*numKeys/capacity);
    fprintf(stderr, "  probe: hits=%zu, misses=%zu, probeRows=%zu\n", probeHits, probeMisses, numProbeRows);
    fprintf(stderr, "  final fill: %zu/%zu = %.1f%% (expand threshold=90%%)\n",
            distinctKeys, capacity, 100.0*distinctKeys/capacity);

    std::mt19937_64 rng(SEED);
    TestData d;
    d.numKeys = numKeys;
    d.numChunks = numChunks;

    d.strCols.resize(NUM_STR_COLS);
    for (size_t c = 0; c < NUM_STR_COLS; c++) {
        d.strCols[c].resize(numKeys);
        for (size_t i = 0; i < numKeys; i++) {
            auto s = "key_" + std::to_string(i) + "_c" + std::to_string(c);
            d.strCols[c][i].assign(s.begin(), s.end());
        }
    }
    std::vector<int64_t> bh(numKeys);
    for (size_t i = 0; i < numKeys; i++) {
        uint64_t h = 0;
        for (size_t c = 0; c < NUM_STR_COLS; c++)
            h = HB(d.strCols[c][i].data(), d.strCols[c][i].size(), h);
        bh[i] = static_cast<int64_t>(h);
    }

    std::vector<std::vector<std::vector<uint8_t>>> ps(NUM_STR_COLS);
    std::vector<int64_t> ph;
    for (size_t i = 0; i < probeHits; i++) {
        size_t idx = rng() % numKeys;
        for (size_t c = 0; c < NUM_STR_COLS; c++) ps[c].push_back(d.strCols[c][idx]);
        ph.push_back(bh[idx]);
    }
    for (size_t i = 0; i < probeMisses; i++) {
        uint64_t h = 0;
        for (size_t c = 0; c < NUM_STR_COLS; c++) {
            auto s = "miss_" + std::to_string(i) + "_" + std::to_string(c);
            std::vector<uint8_t> b(s.begin(), s.end());
            h = HB(b.data(), b.size(), h);
            ps[c].push_back(std::move(b));
        }
        ph.push_back(static_cast<int64_t>(h));
    }

    std::vector<size_t> ord(numProbeRows);
    std::iota(ord.begin(), ord.end(), 0);
    for (size_t i = numProbeRows - 1; i > 0; i--) std::swap(ord[i], ord[rng() % (i + 1)]);
    for (size_t c = 0; c < NUM_STR_COLS; c++) {
        auto tmp = std::move(ps[c]); ps[c].resize(numProbeRows);
        for (size_t i = 0; i < numProbeRows; i++) ps[c][i] = std::move(tmp[ord[i]]);
    }
    { auto tmp = ph; for (size_t i = 0; i < numProbeRows; i++) ph[i] = tmp[ord[i]]; }

    d.totalRows = numKeys + numProbeRows;
    for (size_t c = 0; c < NUM_STR_COLS; c++) {
        d.strCols[c].reserve(d.totalRows);
        for (auto& v : ps[c]) d.strCols[c].push_back(std::move(v));
    }
    d.hashes = bh;
    d.hashes.insert(d.hashes.end(), ph.begin(), ph.end());
    d.values.resize(d.totalRows);
    for (size_t i = 0; i < d.totalRows; i++) d.values[i] = i % 1000;

    d.slices.resize(NUM_STR_COLS);
    for (size_t c = 0; c < NUM_STR_COLS; c++) {
        d.slices[c].resize(d.totalRows);
        for (size_t i = 0; i < d.totalRows; i++) {
            d.slices[c][i].ptr = d.strCols[c][i].data();
            d.slices[c][i].len = d.strCols[c][i].size();
        }
    }

    return d;
}

// ═══════════════════════════════════════════════════════════════════
// Step benchmarks
// ═══════════════════════════════════════════════════════════════════

// 1. Precompute positions: hash & mask
__attribute__((noinline))
static uint64_t BenchPrecomputePositions(const TestData& d) {
    uint32_t mask = static_cast<uint32_t>(d.numChunks - 1);
    uint64_t checksum = 0;
    for (size_t i = 0; i < d.totalRows; i++) {
        uint32_t pos = static_cast<uint32_t>(d.hashes[i]) & mask;
        checksum += pos;
    }
    return checksum;
}

// 5. NewRow
__attribute__((noinline))
static uint64_t BenchNewRow(const TestData& d) {
    taper::SimpleArenaAllocator pool;
    std::vector<size_t> ks(4, 0);
    std::vector<taper::ColumnKind> kinds(4, taper::ColumnKind::Varchar);
    taper::RowContainer rc(ks, kinds, 8, pool);
    uint64_t checksum = 0;
    for (size_t i = 0; i < d.totalRows; i++) {
        char* row = rc.NewRow();
        checksum += reinterpret_cast<uint64_t>(row);
    }
    return checksum;
}

// 7. StoreKeyOneRow (serialize 4 varchar to arena)
__attribute__((noinline, flatten))
static uint64_t BenchSerialize(const TestData& d) {
    taper::SimpleArenaAllocator pool;
    uint64_t checksum = 0;
    for (size_t i = 0; i < d.totalRows; i++) {
        // Manually unroll: compute total size for 4 columns
        size_t len0 = d.slices[0][i].len, len1 = d.slices[1][i].len;
        size_t len2 = d.slices[2][i].len, len3 = d.slices[3][i].len;
        size_t totalSize = (1 + taper::ComputeRowLenSize(len0) + len0)
                         + (1 + taper::ComputeRowLenSize(len1) + len1)
                         + (1 + taper::ComputeRowLenSize(len2) + len2)
                         + (1 + taper::ComputeRowLenSize(len3) + len3);
        uint8_t* block = pool.Allocate(static_cast<int64_t>(totalSize));
        uint8_t* wp = block;
        // Manually unroll: serialize 4 columns
        wp += taper::SerializeVarcharToBuffer(wp, d.slices[0][i].ptr, len0);
        wp += taper::SerializeVarcharToBuffer(wp, d.slices[1][i].ptr, len1);
        wp += taper::SerializeVarcharToBuffer(wp, d.slices[2][i].ptr, len2);
        wp += taper::SerializeVarcharToBuffer(wp, d.slices[3][i].ptr, len3);
        checksum += reinterpret_cast<uint64_t>(block);
    }
    return checksum;
}

// 8. StoreValue<i64>
__attribute__((noinline))
static uint64_t BenchStoreValue(const TestData& d) {
    // Allocate rows first
    taper::SimpleArenaAllocator pool;
    std::vector<size_t> ks(4, 0);
    std::vector<taper::ColumnKind> kinds(4, taper::ColumnKind::Varchar);
    taper::RowContainer rc(ks, kinds, 8, pool);
    std::vector<char*> rows(d.totalRows);
    for (size_t i = 0; i < d.totalRows; i++) rows[i] = rc.NewRow();

    int32_t aggOffset = rc.AggStateOffset();
    uint64_t checksum = 0;
    for (size_t i = 0; i < d.totalRows; i++) {
        taper::RowContainer::StoreValue<int64_t>(rows[i], aggOffset, d.values[i]);
        checksum += static_cast<uint64_t>(d.values[i]);
    }
    return checksum;
}

// 10. BatchCompareVarchar (compare 4 cols, all equal)
__attribute__((noinline, flatten))
static uint64_t BenchCompareVarchar(const TestData& d) {
    // Pre-serialize all rows (unrolled)
    taper::SimpleArenaAllocator pool;
    std::vector<const uint8_t*> blocks(d.totalRows);
    for (size_t i = 0; i < d.totalRows; i++) {
        size_t len0 = d.slices[0][i].len, len1 = d.slices[1][i].len;
        size_t len2 = d.slices[2][i].len, len3 = d.slices[3][i].len;
        size_t totalSize = (1 + taper::ComputeRowLenSize(len0) + len0)
                         + (1 + taper::ComputeRowLenSize(len1) + len1)
                         + (1 + taper::ComputeRowLenSize(len2) + len2)
                         + (1 + taper::ComputeRowLenSize(len3) + len3);
        uint8_t* block = pool.Allocate(static_cast<int64_t>(totalSize));
        uint8_t* wp = block;
        wp += taper::SerializeVarcharToBuffer(wp, d.slices[0][i].ptr, len0);
        wp += taper::SerializeVarcharToBuffer(wp, d.slices[1][i].ptr, len1);
        wp += taper::SerializeVarcharToBuffer(wp, d.slices[2][i].ptr, len2);
        wp += taper::SerializeVarcharToBuffer(wp, d.slices[3][i].ptr, len3);
        blocks[i] = block;
    }

    // Compare (100% equal, unrolled)
    uint64_t match_count = 0;
    for (size_t i = 0; i < d.totalRows; i++) {
        const uint8_t* pos = blocks[i];
        if (!taper::CompareVarcharFromRow(pos, d.slices[0][i].ptr, d.slices[0][i].len)) continue;
        pos += taper::ComputeVarCharSerializedSize(pos);
        if (!taper::CompareVarcharFromRow(pos, d.slices[1][i].ptr, d.slices[1][i].len)) continue;
        pos += taper::ComputeVarCharSerializedSize(pos);
        if (!taper::CompareVarcharFromRow(pos, d.slices[2][i].ptr, d.slices[2][i].len)) continue;
        pos += taper::ComputeVarCharSerializedSize(pos);
        if (!taper::CompareVarcharFromRow(pos, d.slices[3][i].ptr, d.slices[3][i].len)) continue;
        match_count++;
    }
    return match_count;
}

// 12. Accumulate equals: agg += value
__attribute__((noinline))
static uint64_t BenchAccumulate(const TestData& d) {
    taper::SimpleArenaAllocator pool;
    std::vector<size_t> ks(4, 0);
    std::vector<taper::ColumnKind> kinds(4, taper::ColumnKind::Varchar);
    taper::RowContainer rc(ks, kinds, 8, pool);
    int32_t aggOffset = rc.AggStateOffset();
    std::vector<char*> rows(d.totalRows);
    for (size_t i = 0; i < d.totalRows; i++) {
        rows[i] = rc.NewRow();
        taper::RowContainer::StoreValue<int64_t>(rows[i], aggOffset, 0);
    }

    uint64_t checksum = 0;
    for (size_t i = 0; i < d.totalRows; i++) {
        int64_t* p = reinterpret_cast<int64_t*>(rows[i % d.numKeys] + aggOffset);
        *p += d.values[i];
        checksum += static_cast<uint64_t>(*p);
    }
    return checksum;
}

// Full pipeline (for comparison)
__attribute__((noinline, flatten))
static uint64_t BenchFullPipeline(const TestData& d) {
    taper::SimpleArenaAllocator pool;
    std::vector<taper::ColumnDesc> cd(NUM_STR_COLS, taper::ColumnDesc::Varchar);
    taper::TaperColumnSerializeHandler t(pool, 8, cd, d.numChunks);
    size_t numBatches = (d.totalRows + BATCH_SIZE - 1) / BATCH_SIZE;
    std::vector<taper::ColumnInput> cols(NUM_STR_COLS);
    for (size_t batch = 0; batch < numBatches; batch++) {
        size_t start = batch * BATCH_SIZE;
        size_t end = std::min(start + BATCH_SIZE, d.totalRows);
        int32_t batchLen = static_cast<int32_t>(end - start);
        for (size_t c = 0; c < NUM_STR_COLS; c++)
            cols[c] = taper::ColumnInput::MakeVarchar(d.slices[c].data() + start);
        t.EmplaceTableWithDecode(d.hashes.data() + start, batchLen, cols, d.values.data() + start);
    }
    return t.NumGroups();
}

// ═══════════════════════════════════════════════════════════════════
// Usage: ./cpp_step_by_step <sel> [iters] [ht_size] [load_factor]
int main(int argc, char** argv) {
    double sel = 0.1;
    size_t numIters = DEFAULT_ITERS;
    if (argc > 1) sel = std::atof(argv[1]);
    if (argc > 2) numIters = static_cast<size_t>(std::atoi(argv[2]));
    if (argc > 3) G_HT_SIZE = static_cast<size_t>(std::atoi(argv[3]));

    fprintf(stderr, "=== C++ Step-by-Step Bench ===\n");
    fprintf(stderr, "sel=%.2f, iters=%zu, ht=%zu\n", sel, numIters, G_HT_SIZE);
    fprintf(stderr, "Generating data...\n");
    TestData data = GenData(sel);
    fprintf(stderr, "totalRows=%zu, numKeys=%zu, numChunks=%zu\n\n", data.totalRows, data.numKeys, data.numChunks);

    auto bench = [&](const char* name, uint64_t(*fn)(const TestData&)) {
        volatile uint64_t w = fn(data); (void)w;
        auto t0 = Clock::now();
        volatile uint64_t checksum = 0;
        for (size_t i = 0; i < numIters; i++) checksum = fn(data);
        auto t1 = Clock::now();
        double per_iter = elapsed_ms(t0, t1) / numIters;
        printf("%-30s  %7.2f ms  checksum=%lu\n", name, per_iter, (unsigned long)checksum);
    };

    printf("=== C++ Step-by-Step (ht=%zu, sel=%.2f, %zu iters, %zu rows) ===\n", G_HT_SIZE, sel, numIters, data.totalRows);
    bench("1. precompute_positions", BenchPrecomputePositions);
    bench("5. new_row", BenchNewRow);
    bench("7. serialize_4str", BenchSerialize);
    bench("8. store_value_i64", BenchStoreValue);
    bench("10. compare_varchar_4col", BenchCompareVarchar);
    bench("12. accumulate_equals", BenchAccumulate);
    bench("FULL: pipeline", BenchFullPipeline);

    return 0;
}

File: rust/src/bin/step_by_step_bench.rs
//! Step-by-step function benchmark — tests every function in the EmplaceTableWithDecode pipeline.
//!
//! Usage: ./step_by_step_bench <sel> [iterations]
//! Default: sel=0.1, 10 iterations, 1M rows per iteration

use std::time::Instant;
use taper_hashmap::column_marshaller::{
    TaperColumnSerializeHandler, ColumnDesc, ColumnInput,
    serialize_varchar_to_buffer, compute_row_len_size, compare_varchar_from_row,
    compute_varchar_serialized_size,
};
use taper_hashmap::row_container::{RowContainer, ColumnKind};
use xxhash_rust::xxh3::xxh3_64_with_seed;
use rand_mt::Mt19937GenRand64;

const NUM_STR_COLS: usize = 4;
const NUM_PROBE_ROWS: usize = 1_000_000;
const BATCH_SIZE: usize = 410;
const SEED: u64 = 42;
const DEFAULT_ITERS: usize = 10;

fn hash_bytes(data: &[u8], seed: u64) -> u64 { xxh3_64_with_seed(data, seed) }
fn gen_string(base: &str, id: usize, col: usize) -> Vec<u8> { format!("{}_{}_c{}", base, id, col).into_bytes() }

struct TestData {
    str_cols: Vec<Vec<Vec<u8>>>,
    hashes: Vec<u64>,
    values: Vec<i64>,
    total_rows: usize,
    num_keys: usize,
    num_chunks: usize,
}

fn gen_data(sel: f64, ht_size: usize, load_factor: f64) -> TestData {
    let num_keys = (ht_size as f64 * load_factor) as usize;
    let mut rng = Mt19937GenRand64::new(SEED);
    let mut str_cols: Vec<Vec<Vec<u8>>> = (0..NUM_STR_COLS)
        .map(|c| (0..num_keys).map(|i| gen_string("key", i, c)).collect()).collect();
    let build_hashes: Vec<u64> = (0..num_keys).map(|i| {
        let mut h = 0u64;
        for c in 0..NUM_STR_COLS { h = hash_bytes(&str_cols[c][i], h); }
        h
    }).collect();

    let num_hits = (NUM_PROBE_ROWS as f64 * sel) as usize;
    let num_misses = NUM_PROBE_ROWS - num_hits;
    let mut probe_str: Vec<Vec<Vec<u8>>> = vec![Vec::new(); NUM_STR_COLS];
    let mut probe_hashes: Vec<u64> = Vec::new();
    for _ in 0..num_hits { let idx = (rng.next_u64() as usize) % num_keys; for c in 0..NUM_STR_COLS { probe_str[c].push(str_cols[c][idx].clone()); } probe_hashes.push(build_hashes[idx]); }
    for i in 0..num_misses { let mut h = 0u64; for c in 0..NUM_STR_COLS { let s = format!("miss_{}_{}", i, c).into_bytes(); h = hash_bytes(&s, h); probe_str[c].push(s); } probe_hashes.push(h); }

    let mut order: Vec<usize> = (0..NUM_PROBE_ROWS).collect();
    for i in (1..NUM_PROBE_ROWS).rev() { order.swap(i, (rng.next_u64() as usize) % (i + 1)); }
    let probe_str: Vec<Vec<Vec<u8>>> = (0..NUM_STR_COLS).map(|c| order.iter().map(|&i| probe_str[c][i].clone()).collect()).collect();
    let probe_hashes: Vec<u64> = order.iter().map(|&i| probe_hashes[i]).collect();

    let total_rows = num_keys + NUM_PROBE_ROWS;
    for c in 0..NUM_STR_COLS { str_cols[c].extend(probe_str[c].iter().cloned()); }
    let mut all_hashes = build_hashes; all_hashes.extend_from_slice(&probe_hashes);
    let values: Vec<i64> = (0..total_rows).map(|i| (i % 1000) as i64).collect();

    let distinct_keys = num_keys + num_misses;
    let min_slots = ((distinct_keys as f64 / 0.85) as usize).max(8);
    let num_chunks = ((min_slots + 7) / 8).next_power_of_two();

    TestData { str_cols, hashes: all_hashes, values, total_rows, num_keys, num_chunks }
}

// ═══════════════════════════════════════════════════════════════════

#[inline(never)]
fn bench_precompute_positions(d: &TestData) -> u64 {
    let mask = d.num_chunks - 1;
    let mut checksum: u64 = 0;
    for i in 0..d.total_rows {
        let pos = (d.hashes[i] as usize) & mask;
        checksum = checksum.wrapping_add(pos as u64);
    }
    checksum
}

#[inline(never)]
fn bench_new_row(d: &TestData) -> u64 {
    let ks = vec![0usize; 4];
    let kinds = vec![ColumnKind::Varchar; 4];
    let mut rc = RowContainer::with_kinds(&ks, &kinds, 8);
    let mut checksum: u64 = 0;
    for _ in 0..d.total_rows {
        let row = rc.new_row();
        checksum = checksum.wrapping_add(row as u64);
    }
    checksum
}

#[inline(never)]
fn bench_serialize(d: &TestData) -> u64 {
    let ks = vec![0usize; 4];
    let kinds = vec![ColumnKind::Varchar; 4];
    let mut rc = RowContainer::with_kinds(&ks, &kinds, 8);
    let mut checksum: u64 = 0;
    for i in 0..d.total_rows {
        let mut total_size = 0usize;
        for c in 0..NUM_STR_COLS {
            let s = &d.str_cols[c][i];
            total_size += 1 + compute_row_len_size(s.len()) as usize + s.len();
        }
        let block = rc.arena_alloc(total_size);
        let mut wp = block;
        for c in 0..NUM_STR_COLS {
            let s = &d.str_cols[c][i];
            let written = serialize_varchar_to_buffer(wp, s.as_slice());
            wp = unsafe { wp.add(written) };
        }
        checksum = checksum.wrapping_add(block as u64);
    }
    checksum
}

#[inline(never)]
fn bench_store_value(d: &TestData) -> u64 {
    let ks = vec![0usize; 4];
    let kinds = vec![ColumnKind::Varchar; 4];
    let mut rc = RowContainer::with_kinds(&ks, &kinds, 8);
    let agg_offset = rc.agg_state_offset();
    let mut rows: Vec<*mut u8> = Vec::with_capacity(d.total_rows);
    for _ in 0..d.total_rows { rows.push(rc.new_row()); }

    let mut checksum: u64 = 0;
    for i in 0..d.total_rows {
        RowContainer::store_value::<i64>(rows[i], agg_offset, d.values[i]);
        checksum = checksum.wrapping_add(d.values[i] as u64);
    }
    checksum
}

#[inline(never)]
fn bench_compare_varchar(d: &TestData) -> u64 {
    // Pre-serialize
    let ks = vec![0usize; 4];
    let kinds = vec![ColumnKind::Varchar; 4];
    let mut rc = RowContainer::with_kinds(&ks, &kinds, 8);
    let mut blocks: Vec<*const u8> = Vec::with_capacity(d.total_rows);
    for i in 0..d.total_rows {
        let mut total_size = 0usize;
        for c in 0..NUM_STR_COLS {
            let s = &d.str_cols[c][i];
            total_size += 1 + compute_row_len_size(s.len()) as usize + s.len();
        }
        let block = rc.arena_alloc(total_size);
        let mut wp = block;
        for c in 0..NUM_STR_COLS {
            let s = &d.str_cols[c][i];
            let written = serialize_varchar_to_buffer(wp, s.as_slice());
            wp = unsafe { wp.add(written) };
        }
        blocks.push(block as *const u8);
    }

    // Compare (100% equal)
    let mut match_count: u64 = 0;
    for i in 0..d.total_rows {
        let mut pos = blocks[i];
        let mut all_match = true;
        for c in 0..NUM_STR_COLS {
            let s = &d.str_cols[c][i];
            if !compare_varchar_from_row(pos, s.as_slice()) { all_match = false; break; }
            let entry_size = compute_varchar_serialized_size(pos);
            pos = unsafe { pos.add(entry_size) };
        }
        if all_match { match_count += 1; }
    }
    match_count
}

#[inline(never)]
fn bench_accumulate(d: &TestData) -> u64 {
    let ks = vec![0usize; 4];
    let kinds = vec![ColumnKind::Varchar; 4];
    let mut rc = RowContainer::with_kinds(&ks, &kinds, 8);
    let agg_offset = rc.agg_state_offset();
    let mut rows: Vec<*mut u8> = Vec::with_capacity(d.num_keys);
    for _ in 0..d.num_keys {
        let row = rc.new_row();
        RowContainer::store_value::<i64>(row, agg_offset, 0i64);
        rows.push(row);
    }

    let mut checksum: u64 = 0;
    for i in 0..d.total_rows {
        let row = rows[i % d.num_keys];
        let p = unsafe { &mut *(row.add(agg_offset) as *mut i64) };
        *p += d.values[i];
        checksum = checksum.wrapping_add(*p as u64);
    }
    checksum
}

#[inline(never)]
fn bench_full_pipeline(d: &TestData) -> u64 {
    let col_descs = vec![ColumnDesc::Varchar; NUM_STR_COLS];
    let mut table = TaperColumnSerializeHandler::new(&col_descs, 8, d.num_chunks);
    let num_batches = (d.total_rows + BATCH_SIZE - 1) / BATCH_SIZE;
    for batch_idx in 0..num_batches {
        let start = batch_idx * BATCH_SIZE;
        let end = (start + BATCH_SIZE).min(d.total_rows);
        let batch_hashes = &d.hashes[start..end];
        let batch_values = &d.values[start..end];
        let str_slices: Vec<Vec<&[u8]>> = (0..NUM_STR_COLS)
            .map(|c| d.str_cols[c][start..end].iter().map(|s| s.as_slice()).collect()).collect();
        let mut columns: Vec<ColumnInput> = Vec::new();
        for c in 0..NUM_STR_COLS { columns.push(ColumnInput::Varchar(&str_slices[c])); }
        table.emplace_table_with_decode(batch_hashes, &columns, batch_values);
    }
    table.num_groups() as u64
}

// ═══════════════════════════════════════════════════════════════════
// Usage: ./step_by_step_bench <sel> [iters] [ht_size] [load_factor]
fn main() {
    let args: Vec<String> = std::env::args().collect();
    let sel: f64 = args.get(1).and_then(|s| s.parse().ok()).unwrap_or(0.1);
    let num_iters: usize = args.get(2).and_then(|s| s.parse().ok()).unwrap_or(DEFAULT_ITERS);
    let ht_size: usize = args.get(3).and_then(|s| s.parse().ok()).unwrap_or(16384);
    let load_factor: f64 = args.get(4).and_then(|s| s.parse().ok()).unwrap_or(0.5);

    eprintln!("=== Rust Step-by-Step Bench ===");
    eprintln!("sel={:.2}, iters={}, ht={}, lf={:.2}", sel, num_iters, ht_size, load_factor);
    eprintln!("Generating data...");
    let data = gen_data(sel, ht_size, load_factor);
    eprintln!("totalRows={}, numKeys={}, numChunks={}\n", data.total_rows, data.num_keys, data.num_chunks);

    let bench = |name: &str, f: fn(&TestData) -> u64| {
        let _ = f(&data); // warmup
        let t0 = Instant::now();
        let mut checksum: u64 = 0;
        for _ in 0..num_iters { checksum = f(&data); }
        let per_iter = t0.elapsed().as_secs_f64() * 1000.0 / num_iters as f64;
        println!("{:<30}  {:7.2} ms  checksum={}", name, per_iter, checksum);
    };

    println!("=== Rust Step-by-Step (ht={}, lf={:.2}, sel={:.2}, {} iters, {} rows) ===", ht_size, load_factor, sel, num_iters, data.total_rows);
    bench("1. precompute_positions", bench_precompute_positions);
    bench("5. new_row", bench_new_row);
    bench("7. serialize_4str", bench_serialize);
    bench("8. store_value_i64", bench_store_value);
    bench("10. compare_varchar_4col", bench_compare_varchar);
    bench("12. accumulate_equals", bench_accumulate);
    bench("FULL: pipeline", bench_full_pipeline);
}
